pkg/dns/aws/health.go

```diff
@@ -0,0 +1,252 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1alpha1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+	"github.com/aws/aws-sdk-go/service/route53/route53iface"
+	"github.com/rs/xid"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+const (
+	idTag = "kuadrant.dev/healthcheck"
+
+	defaultHealthCheckPath             = "/"
+	defaultHealthCheckPort             = 80
+	defaultHealthCheckFailureThreshold = 3
+)
+
+var (
+	callerReference func(id string) *string
+)
+
+type Route53HealthCheckReconciler struct {
+	client route53iface.Route53API
+}
+
+var _ dns.HealthCheckReconciler = &Route53HealthCheckReconciler{}
+
+func NewRoute53HealthCheckReconciler(client route53iface.Route53API) *Route53HealthCheckReconciler {
+	return &Route53HealthCheckReconciler{
+		client: client,
+	}
+}
+
+func (r *Route53HealthCheckReconciler) Reconcile(ctx context.Context, spec dns.HealthCheckSpec, endpoint *v1alpha1.Endpoint) (dns.HealthCheckResult, error) {
+	healthCheck, exists, err := r.findHealthCheck(ctx, endpoint)
```

Comment:
Is this going to the AWS API on every reconcile? Could this lead to us being rate limited?